orangecontrib/spectroscopy/tests/test_preprocess.py

```diff
@@ -343,9 +343,14 @@ def test_unknown_no_propagate(self):
         for i in range(min(len(data), len(data.domain.attributes))):
             data.X[i, i] = np.nan
         for proc in PREPROCESSORS:
+            if hasattr(proc, "reference"):
+                max_nans = 1 + data.X.size - np.count_nonzero(proc.reference)
```

Comment:
A bug. For Absorbance this gives max_nans of 42729, which are of course never found.

The problem seemed strange. The data used for the tests and the input main reference all have absolutely no zeros. Then, I saw that the problematic data is caused by `add_zeros`. :)

Solving this makes quite a mess, forces the same name just for the test to work (I do not not mind the same name though). And sometimes, references could be interpolated and thus their length could change, so the formula for expected number of NaNs is unclear.

Because it is hard to write I criterium, I would try attaching a flag like `skip_unknown_not_propagate` to the preprocessor (not reference!) in `add_edge_case_data_parameter` for `add_zeros` and just skip preprocessors with this flag.

Furthermore, `test_no_infs` should also have `data` with zeros, because they can also cause infs if the situation is reversed. You should also have data when a whole row (or, maybe, also a whole column) is a zero, as summing over it is common. But we can never catch everything like this...

Reply:
Good catch!

Thanks for the suggestions. I'll admit I don't know how to attach the `skip_unknown_no_propagate` flag as you suggest.

I'm also unclear how the data (numerator) having zeros can result in an inf?

Reply:
Isn't data also denumerator in some Absorbace? ref/data.X?

Reply:
For `add_edge_case_data_parameter`

```
    ...
    for i, d in enumerate(modified):
        kwargs[data_arg_name] = d
        p = class_(*args, **kwargs)
        if i == 5:
            p.skip_something = True
        yield p
```

And then you can check for it in the test with `hasattr`

Reply:
Thank you.

> Isn't data also denumerator in some Absorbace? ref/data.X?

Right :)